stake-pool: CLI QOL improvements

gabrielhicks opened the pull request:
When creating a stake-pool using the CLI, if at any point an instruction fails subsequent commands will also fail. This behavior creates a problem when initializing a stake pool in the CLI with vanity keypairs. This solution aims to break up the `command_create_pool` into iterative steps with checks along the way.

I had changed the deprecated ` use spl_associated_token_account::create_associated_token_account` but am happy to change it back.

joncinque commented on stake-pool/cli/src/main.rs:
```diff
@@ -444,50 +567,121 @@ fn command_create_pool(
+    setup_reserve_stake_account(config, &reserve_keypair, reserve_stake_balance, &withdraw_authority)?;
+    setup_mint_account(config, &mint_keypair, mint_account_balance, &withdraw_authority, default_decimals)?;
+    setup_pool_fee_account(config, &mint_keypair.pubkey(), &mut total_rent_free_balances)?;
+    setup_validator_list_account(config, &validator_list_keypair, validator_list_balance, validator_list_size)?;
+
+    let pool_fee_account = get_associated_token_address(&config.manager.pubkey(), &mint_keypair.pubkey());
+
+    initialize_stake_pool(
+        config,
+        &stake_pool_keypair,
+        &validator_list_keypair,
+        &reserve_keypair,
+        &mint_keypair,
+        &pool_fee_account,
+        deposit_authority,
+        epoch_fee,
+        withdrawal_fee,
+        deposit_fee,
+        referral_fee,
+        max_validators,
+        &withdraw_authority,
+    )?;
```
Can we combine `setup_validator_list_account` and `initialize_stake_pool` into the same transaction? Fewer transactions is always best. And although the possibility is really remote, it's technically possible for someone to steal a created-but-not-initialized account, since no data has been written to it yet

joncinque commented on stake-pool/cli/src/main.rs:
```diff
@@ -374,66 +344,219 @@ fn command_create_pool(
+        
+        
+    if let Ok(account) = stake_pool_account_info {
+        let stake_pool_manager = get_stake_pool(&config.rpc_client, &stake_pool_keypair.pubkey()).unwrap().manager;
+        if account.owner == spl_stake_pool::id() {
+            if stake_pool_manager == config.manager.pubkey() {
+                println!("Stake pool account {} already exists and is initialized", stake_pool_keypair.pubkey());
+                return Ok(());
+            } else {
+                println!("Stake pool account {} already exists but is not initialized", stake_pool_keypair.pubkey());
+            }
+        }
```
This could unnecessarily panic -- rather than checking the manager, we can just check that the first byte in the data is equal to `AccountType::StakePool as u8` (if there is any data in the account) https://github.com/solana-labs/solana-program-library/blob/c4d51f20b8722d23b03e94c7075702fca52460dd/stake-pool/program/src/state.rs#L37

joncinque commented on stake-pool/cli/src/main.rs:
```diff
@@ -374,66 +344,219 @@ fn command_create_pool(
+            }
+        }
+    } else {
+        // Create the stake pool account if it doesn't exist
+        let create_instructions = vec![
+            system_instruction::create_account(
+                &config.fee_payer.pubkey(),
+                &stake_pool_keypair.pubkey(),
+                stake_pool_account_lamports,
+                get_packed_len::<StakePool>() as u64,
+                &spl_stake_pool::id(),
+            ),
+        ];
+
+        let mut create_signers = vec![
+            config.fee_payer.as_ref(),
+            stake_pool_keypair,
+        ];
+
+        unique_signers!(create_signers);
+        let create_transaction = checked_transaction_with_signers(config, &create_instructions, &create_signers)?;
+
+        println!(
+            "Creating stake pool account {}",
+            stake_pool_keypair.pubkey(),
+        );
+        send_transaction(config, create_transaction)?;
```
Let's put the creation and initialization of the stake pool in the same transaction (along with the creation of the validator list, as mentioned in another comment) -- more transactions just means more opportunities for things to go wrong

joncinque commented on stake-pool/cli/src/main.rs:
```diff
@@ -2,9 +2,9 @@
-// use instruction::create_associated_token_account once ATA 1.0.5 is released
-#[allow(deprecated)]
-use spl_associated_token_account::create_associated_token_account;
+use spl_associated_token_account::instruction::create_associated_token_account; // once ATA 1.0.5 is released
```
This is great, thanks! Normally I'd ask for it to go in a separate PR, but it's a small enough change. You could even remove the comment
```suggestion
use spl_associated_token_account::instruction::create_associated_token_account;
```

joncinque commented on stake-pool/cli/src/main.rs:
```diff
@@ -2,9 +2,9 @@
-#[allow(deprecated)]
-use spl_associated_token_account::create_associated_token_account;
+use spl_associated_token_account::instruction::create_associated_token_account; // once ATA 1.0.5 is released
+// #[allow(deprecated)]
+// use spl_associated_token_account::create_associated_token_account;
```
nit: Feel free to remove these comments

joncinque commented on stake-pool/cli/src/main.rs:
```diff
@@ -775,11 +969,12 @@ fn add_associated_token_account(
             .unwrap();
 
-        #[allow(deprecated)]
+        // #[allow(deprecated)]
```
nit: Feel free to remove this comment too

joncinque reviewed (commented):
Thanks for the contribution! This will definitely make the experience of creating a stake pool much smoother. Let me know if anything's unclear

gabrielhicks commented:
> Thanks for the contribution! This will definitely make the experience of creating a stake pool much smoother. Let me know if anything's unclear

I believe I have walked through each change and implemented! I appreciate the feedback and am happy to continue iterating if needed.

joncinque commented on stake-pool/cli/src/main.rs:
```diff
@@ -374,120 +340,304 @@ fn command_create_pool(
+    send_transaction(config, transaction)?;
+    Ok(())
+}
+
+fn setup_and_initialize_validator_list_with_stake_pool(
```
This needs a clippy allow according to CI
```suggestion

#[allow(clippy::too_many_arguments)]
fn setup_and_initialize_validator_list_with_stake_pool(
```

joncinque reviewed (commented):
Thanks for addressing all of these! Just a few last comments before merging.

I've kicked off CI, and it's failing the formatting check -- can you run `./cargo-nightly fmt --all` from the top of the repo?

You'll also need to run `./cargo-nightly clippy -Zunstable-options --all-targets -- --deny=warnings --deny=clippy::arithmetic_side_effects` and make sure everything's ok there. I think it'll be OK after addressing the remaining comments.

gabrielhicks commented on stake-pool/cli/src/main.rs:
```diff
@@ -374,120 +340,304 @@ fn command_create_pool(
-        ),
-        // Initialize stake pool
-        spl_stake_pool::instruction::initialize(
+        ));
+        instructions.push(spl_stake_pool::instruction::initialize(
```
I was wondering if I should even have checks here as my assumption and understanding is if you are making it this far, we want you to do all of the ixs together

joncinque replied:
Yeah that's true, but who knows, maybe someone wants to initialize one of the accounts separately themselves? It can't hurt, thankfully

gabrielhicks commented:
> Thanks for addressing all of these! Just a few last comments before merging.
> 
> I've kicked off CI, and it's failing the formatting check -- can you run `./cargo-nightly fmt --all` from the top of the repo?
> 
> You'll also need to run `./cargo-nightly clippy -Zunstable-options --all-targets -- --deny=warnings --deny=clippy::arithmetic_side_effects` and make sure everything's ok there. I think it'll be OK after addressing the remaining comments.

Ran with changes and all looks good! Thank you!

joncinque reviewed (approved):
Thanks again for your contribution!
